deployment/environment/devenv/environment.go

```diff
@@ -63,3 +72,257 @@ func NewEnvironment(ctx func() context.Context, lggr logger.Logger, config Envir
+	}
+	b.nodeSetOutput = nodeSetOutput
+	return b
+}
+
+func (b *EnvironmentBuilder) WithExistingAddresses(existingAddresses deployment.AddressBook) *EnvironmentBuilder {
+	b.existingAddresses = existingAddresses
+	return b
+}
+
+func (b *EnvironmentBuilder) WithTopology(topology *types.Topology) *EnvironmentBuilder {
+	if topology != nil {
+		if len(topology.DonsMetadata) == 0 {
+			b.errs = append(b.errs, "metadata not set")
+		}
+		if topology.WorkflowDONID == 0 {
+			b.errs = append(b.errs, "workflow don id not set")
+		}
+	}
+
+	b.topology = topology
+	return b
+}
+
+func (b *EnvironmentBuilder) Build() (*EnvironmentWithTopology, error) {
+	if len(b.errs) > 0 {
+		return nil, errors.New("validation errors: " + strings.Join(b.errs, ", "))
+	}
+	if b.topology == nil {
+		return nil, errors.New("topology not set")
+	}
+	if b.blockchainOutputs == nil {
+		return nil, errors.New("blockchain outputs not set")
+	}
+	if b.nodeSetOutput == nil {
+		return nil, errors.New("nodeSetOutput not set")
+	}
+	if b.jdOutput == nil {
+		return nil, errors.New("jd output not set")
+	}
+
+	envs := make([]*deployment.Environment, len(b.nodeSetOutput))
+	dons := make([]*DON, len(b.nodeSetOutput))
+
+	var allNodesInfo []NodeInfo
+	chains := make([]ChainConfig, 0)
+
+	for _, sethClient := range b.sethClients {
+		blockchainOutput := b.blockchainOutputs[strconv.FormatInt(sethClient.ChainID, 10)]
+		chainConfig := ChainConfig{
+			ChainID:   sethClient.Cfg.Network.ChainID,
+			ChainName: sethClient.Cfg.Network.Name,
+			ChainType: strings.ToUpper(blockchainOutput.Family),
+			WSRPCs: []CribRPCs{{
+				External: blockchainOutput.Nodes[0].ExternalWSUrl,
+				Internal: blockchainOutput.Nodes[0].InternalWSUrl,
+			}},
+			HTTPRPCs: []CribRPCs{{
+				External: blockchainOutput.Nodes[0].ExternalHTTPUrl,
+				Internal: blockchainOutput.Nodes[0].InternalHTTPUrl,
+			}},
+			DeployerKey: sethClient.NewTXOpts(seth.WithNonce(nil)), // set nonce to nil, so that it will be fetched from the chain
+		}
+		chains = append(chains, chainConfig)
+	}
+
+	for idx, nodeOutput := range b.nodeSetOutput {
+		// check how many bootstrap nodes we have in each DON
+		bootstrapNodes, err := FindManyWithLabel(b.topology.DonsMetadata[idx].NodesMetadata, &types.Label{Key: types.NodeTypeKey, Value: types.BootstrapNode}, EqualLabels)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to find bootstrap nodes")
+		}
+		nodeInfo, err := GetNodeInfo(nodeOutput.Output, nodeOutput.NodeSetName, len(bootstrapNodes))
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get node info")
+		}
+		allNodesInfo = append(allNodesInfo, nodeInfo...)
+
+		// if DON has no capabilities and is not a CCIP DON, we don't need to create chain configs (e.g. for gateway nodes)
+		// we indicate to `NewEnvironment` that it should skip chain creation by passing an empty chain config
+		if len(nodeOutput.Capabilities) == 0 && nodeOutput.NodeSetType != "ccip" {
```

Comment:
I worry this is brittle, what if someone names it differently? Also I am not sure we should mix CRE and non-CRE together. CCIP doesn't use CRE and might never use it. Why mix both things?